CrashTracer: com.apple.WebKit.WebContent at CoreGraphics: CGDataProviderCopyData

grorg opened the pull request:
#### a5aad74576d8c44d3fff0d5b50bbffc485e6ac62
<pre>
CrashTracer: com.apple.WebKit.WebContent at CoreGraphics: CGDataProviderCopyData
<a href="https://bugs.webkit.org/show_bug.cgi?id=259296">https://bugs.webkit.org/show_bug.cgi?id=259296</a>
rdar://29394886

Reviewed by Simon Fraser.

The change in <a href="https://commits.webkit.org/262607@main">https://commits.webkit.org/262607@main</a> has caused an
occasional out of memory exception to be thrown in CGDataProviderCopyData.
Catch and ignore this exception, while bailing out of the method.

I&apos;ve added some logging to give more information if we see it again.

This involved changing a .cpp file to .mm.

* Source/WebKit/Shared/cg/ShareableBitmapCG.mm: Renamed from Source/WebKit/Shared/cg/ShareableBitmapCG.cpp.
(WebKit::ShareableBitmapConfiguration::ShareableBitmapConfiguration):
(WebKit::ShareableBitmapConfiguration::validateColorSpace):
(WebKit::wantsExtendedRange):
(WebKit::ShareableBitmapConfiguration::calculateBytesPerPixel):
(WebKit::ShareableBitmapConfiguration::calculateBytesPerRow):
(WebKit::ShareableBitmapConfiguration::calculateBitmapInfo):
(WebKit::ShareableBitmap::createFromImagePixels):
(WebKit::ShareableBitmap::createGraphicsContext):
(WebKit::ShareableBitmap::paint):
(WebKit::ShareableBitmap::makeCGImageCopy):
(WebKit::ShareableBitmap::makeCGImage):
(WebKit::ShareableBitmap::createPlatformImage):
(WebKit::ShareableBitmap::createCGImage const):
(WebKit::ShareableBitmap::releaseBitmapContextData):
(WebKit::ShareableBitmap::createImage):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/266119@main">https://commits.webkit.org/266119@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4ce88c1e2544055306fa92db36f87dda2627dbf1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12941 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13264 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13595 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14679 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/12350 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15769 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13286 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15044 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13107 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13810 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10948 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15130 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11098 "Passed tests") | | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/18765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12173 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11868 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15074 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12336 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/10229 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11598 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3169 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15917 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12177 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/Shared/cg/ShareableBitmapCG.mm:
```diff
@@ -115,7 +115,20 @@ RefPtr<ShareableBitmap> ShareableBitmap::createFromImagePixels(NativeImage& imag
+    @try {
+        pixels = adoptCF(CGDataProviderCopyData(CGImageGetDataProvider(image.platformImage().get())));
+    } @catch (id exception) {
+        LOG_WITH_STREAM(Images, stream
```
Log the image URL if you can.

grorg replied:
I can't see a way to get from NativeImage to the originating URL.

grorg committed:
CrashTracer: com.apple.WebKit.WebContent at CoreGraphics: CGDataProviderCopyData
https://bugs.webkit.org/show_bug.cgi?id=259296
rdar://29394886

Reviewed by Simon Fraser.

The change in https://commits.webkit.org/262607@main has caused an
occasional out of memory exception to be thrown in CGDataProviderCopyData.
Catch and ignore this exception, while bailing out of the method.

I've added some logging to give more information if we see it again.

This involved changing a .cpp file to .mm.

* Source/WebKit/Shared/cg/ShareableBitmapCG.mm: Renamed from Source/WebKit/Shared/cg/ShareableBitmapCG.cpp.
(WebKit::ShareableBitmapConfiguration::ShareableBitmapConfiguration):
(WebKit::ShareableBitmapConfiguration::validateColorSpace):
(WebKit::wantsExtendedRange):
(WebKit::ShareableBitmapConfiguration::calculateBytesPerPixel):
(WebKit::ShareableBitmapConfiguration::calculateBytesPerRow):
(WebKit::ShareableBitmapConfiguration::calculateBitmapInfo):
(WebKit::ShareableBitmap::createFromImagePixels):
(WebKit::ShareableBitmap::createGraphicsContext):
(WebKit::ShareableBitmap::paint):
(WebKit::ShareableBitmap::makeCGImageCopy):
(WebKit::ShareableBitmap::makeCGImage):
(WebKit::ShareableBitmap::createPlatformImage):
(WebKit::ShareableBitmap::createCGImage const):
(WebKit::ShareableBitmap::releaseBitmapContextData):
(WebKit::ShareableBitmap::createImage):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/266119@main
